### PR TITLE
More explicitly specify the algorithm of encode/decode methods

### PR DESCRIPTION
**Pull request by Josh-Cena**
### Description

These reference pages are too handwavy and don't really discuss what happens with all kinds of inputs. What characters are encoded/decoded are too hard to follow.

### CODE REVIEW

File: files/en-us/web/javascript/reference/global_objects/decodeuri/index.md [hunk 1]
```diff
@@ -35,32 +33,52 @@ decodeURI(encodedURI)
...
 
 ## Description
 
-Replaces each escape sequence in the encoded URI with the character that it represents,
-but does not decode escape sequences that could not have been introduced by
-{{jsxref("encodeURI")}}. The character `#` is not decoded from escape
-sequences.
+`decodeURI()` is a function property of the global object.
+
+The `decodeURI()` function decodes characters by UTF-8 code units. In UTF-8, the number of leading 1 bits in thw first byte, which may be 0 (for ASCII characters), 2, 3, or 4, indicates the number of bytes (each encoded in the form `%XX`) in the sequence. So by reading the first escape sequence, `decodeURI()` can determine how many more escape sequences to consume. If `decodeURI()` fails to find the expected number of sequences, it throws a {{jsxref("URIError")}}.
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/decodeuri/index.md:47]: Fly by typo
```suggestion
The `decodeURI()` function decodes characters by UTF-8 code units. In UTF-8, the number of leading 1 bits in the first byte, which may be 0 (for ASCII characters), 2, 3, or 4, indicates the number of bytes (each encoded in the form `%XX`) in the sequence. So by reading the first escape sequence, `decodeURI()` can determine how many more escape sequences to consume. If `decodeURI()` fails to find the expected number of sequences, it throws a {{jsxref("URIError")}}.
```

This sentence is hard to read. YOu might think about splitting it up a bit.

File: files/en-us/web/javascript/reference/global_objects/escape/index.md [hunk 2]
```diff
@@ -53,17 +32,15 @@ A new string in which certain characters have been escaped.
...
-characters are encoded with the exception of: `@*_+-./`
+`escape()` is a function property of the global object.
+
+The `escape()` function replaces all characters with a escape sequence, with the exception of ASCII word characters (A–Z, a–z, 0–9, _) and `@*_+-./`. Characters are escaped by UTF-16 code units. If the code unit's value is less than 256, it is represented by a two-digit hexadecimal number in the format `%XX`, left-padded with 0 if necessary. Otherwise, it is represented by a four-digit hexadecimal number in the format `%uXXXX`, left-padded with 0 if necessary.
 
-The hexadecimal form for characters, whose code unit value is `0xFF` or
-less, is a two-digit escape sequence: `%xx`. For characters with a
-greater code unit, the four-digit format `%uxxxx`
-is used.
+> **Note:** This function was used mostly for URL encoding and is partly based on the escape format in {{rfc(1738)}}. The escape format is _not_ a [escape sequence](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#escape_sequences) in string literals. You can replace `%XX` with `\xXX` and `%uXXXX` with `\uXXXX` to get a string containing actual string literal escape sequences.
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/reference/global_objects/escape/index.md:39]:
```suggestion
> **Note:** This function was used mostly for URL encoding and is partly based on the escape format in {{rfc(1738)}}. The escape format is _not_ an [escape sequence](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#escape_sequences) in string literals. You can replace `%XX` with `\xXX` and `%uXXXX` with `\uXXXX` to get a string containing actual string-literal escape sequences.
```

File: files/en-us/web/javascript/reference/global_objects/unescape/index.md [hunk 3]
```diff
@@ -11,25 +11,9 @@ browser-compat: javascript.builtins.unescape
...
-hexadecimal escape sequences are replaced with the character that it represents. The
-escape sequences might be introduced by a function like {{jsxref("escape")}}. Usually,
-{{jsxref("decodeURI")}} or {{jsxref("decodeURIComponent")}} are preferred over
-`unescape`.
-
-> **Note:** Do not use `unescape` to decode URIs,
-> use `decodeURI` instead.
+> **Note:** `unescape()` is a non-standard function implemented by browsers and was only standardized for cross-engine compatibility. It is not required to be implemented by all JavaScript engines and may not work everywhere. Use {{jsxref("decodeURIComponent()")}} or {{jsxref("decodeURI()")}} if possible.
+
+The **`unescape()`** function computes a new string in which hexadecimal escape sequences are replaced with the character that it represents. The escape sequences might be introduced by a function like {{jsxref("escape()")}}.
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/reference/global_objects/unescape/index.md:16]:
```suggestion
The **`unescape()`** function computes a new string in which hexadecimal escape sequences are replaced with the characters that they represent. The escape sequences might be introduced by a function like {{jsxref("escape()")}}.
```

File: files/en-us/web/javascript/reference/global_objects/unescape/index.md [hunk 4]
```diff
@@ -48,11 +32,15 @@ A new string in which certain characters have been unescaped.
 
 ## Description
 
-The `unescape` function is a property of the _global object_.
+`unescape()` is a function property of the global object.
+
+The `unescape()` function replaces all escape sequences with the character that it represents. Specifically, it replaces all escape sequences of the form `%XX` or `%uXXXX` (where `X` represents one hexadecimal digit) with the character that has the hexadecimal value `XX`/`XXXX`. If the escape sequence is not a valid escape sequence (for example, if `%` is followed by one or no hex digit), it is left as-is.
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/reference/global_objects/unescape/index.md:37]:
```suggestion
The `unescape()` function replaces any escape sequence with the character that it represents. Specifically, it replaces any escape sequence of the form `%XX` or `%uXXXX` (where `X` represents one hexadecimal digit) with the character that has the hexadecimal value `XX`/`XXXX`. If the escape sequence is not a valid escape sequence (for example, if `%` is followed by one or no hex digit), it is left as-is.
```

File: files/en-us/web/javascript/reference/global_objects/escape/index.md [hunk 5]
```diff
@@ -53,17 +32,15 @@ A new string in which certain characters have been escaped.
 
 ## Description
 
-The `escape` function is a property of the _global object_. Special
-characters are encoded with the exception of: `@*_+-./`
+`escape()` is a function property of the global object.
+
+The `escape()` function replaces all characters with a escape sequence, with the exception of ASCII word characters (A–Z, a–z, 0–9, _) and `@*_+-./`. Characters are escaped by UTF-16 code units. If the code unit's value is less than 256, it is represented by a two-digit hexadecimal number in the format `%XX`, left-padded with 0 if necessary. Otherwise, it is represented by a four-digit hexadecimal number in the format `%uXXXX`, left-padded with 0 if necessary.
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/reference/global_objects/escape/index.md:37]:
```suggestion
The `escape()` function replaces all characters with escape sequences, with the exception of ASCII word characters (A–Z, a–z, 0–9, _) and `@*_+-./`. Characters are escaped by UTF-16 code units. If the code unit's value is less than 256, it is represented by a two-digit hexadecimal number in the format `%XX`, left-padded with 0 if necessary. Otherwise, it is represented by a four-digit hexadecimal number in the format `%uXXXX`, left-padded with 0 if necessary.
```

File: files/en-us/web/javascript/reference/global_objects/escape/index.md [hunk 6]
```diff
@@ -11,30 +11,9 @@ browser-compat: javascript.builtins.escape
...
-> which use the format `\xHH`. (`HH` are
-> two hexadecimal digits, and the form `\xHH\xHH` is
-> used for higher-plane Unicode characters.)
->
-> Escaped characters in String literals can be expanded by replacing the
-> `\x` with `%`, then using the `decodeURIComponent()`
-> function.
+> **Note:** `escape()` is a non-standard function implemented by browsers and was only standardized for cross-engine compatibility. It is not required to be implemented by all JavaScript engines and may not work everywhere. Use {{jsxref("encodeURIComponent()")}} or {{jsxref("encodeURI()")}} if possible.
+
+The **`escape()`** function computes a new string in which certain characters have been replaced by a hexadecimal escape sequence.
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/reference/global_objects/escape/index.md:16]:
```suggestion
The **`escape()`** function computes a new string in which certain characters have been replaced by hexadecimal escape sequences.
```

**Review by sideshowbarker — APPROVED** [review]
